pkg/serviceprovider/quay/metadataprovider.go

```diff
@@ -80,53 +67,138 @@ func (s metadataProvider) Fetch(ctx context.Context, token *api.SPIAccessToken)
-		},
-	})
+type RepositoryMetadata struct {
+	Repository   EntityRecord
+	Organization EntityRecord
+}
+
+// FetchRepo is the iterative version of Fetch used internally in the Quay service provider. It takes care of both
+// fetching and caching of the data on per-repository basis.
+func (p metadataProvider) FetchRepo(ctx context.Context, repoUrl string, token *api.SPIAccessToken) (metadata RepositoryMetadata, err error) {
+	if token.Status.TokenMetadata == nil {
+		return
+	}
+
+	quayState := TokenState{}
+	if err = json.Unmarshal(token.Status.TokenMetadata.ServiceProviderState, &quayState); err != nil {
+		return
+	}
+
+	var tokenData *api.Token
+
+	tokenData, err = p.tokenStorage.Get(ctx, token)
+	if err != nil {
+		return
+	}
+
+	orgOrUser, repo, _ := splitToOrganizationAndRepositoryAndVersion(repoUrl)
+
+	// enable the lazy one-time login to docker in the subsequent calls
+	var loginToken *LoginTokenInfo
+	getLoginTokenInfo := func() (LoginTokenInfo, error) {
+		if loginToken != nil {
+			return *loginToken, nil
+		}
+
+		username, password := getUsernameAndPasswordFromTokenData(tokenData)
+		var loginToken string
+		loginToken, err = DockerLogin(ctx, p.httpClient, repo, username, password)
+		if err != nil {
+			return LoginTokenInfo{}, err
+		}
+
+		info, err := AnalyzeLoginToken(loginToken)
+		if err != nil {
+			return LoginTokenInfo{}, err
+		}
+
+		return info, nil
+	}
+
+	var orgChanged, repoChanged bool
+	var orgRecord, repoRecord EntityRecord
+
+	orgRecord, orgChanged, err = p.getEntityRecord(ctx, token, orgOrUser, quayState.Organizations, getLoginTokenInfo, fetchOrganizationRecord)
 	if err != nil {
 		return
 	}
 
-	content := map[string]interface{}{}
-	if err = json.NewDecoder(res.Body).Decode(&content); err != nil {
+	repoRecord, repoChanged, err = p.getEntityRecord(ctx, token, orgOrUser+"/"+repo, quayState.Repositories, getLoginTokenInfo, fetchRepositoryRecord)
+	if err != nil {
 		return
 	}
 
-	userName = content["username"].(string)
+	if orgChanged || repoChanged {
+		if err = p.persistTokenState(ctx, token, &quayState); err != nil {
+			return
+		}
+	}
 
-	return
+	return RepositoryMetadata{
+		Repository:   repoRecord,
+		Organization: orgRecord,
+	}, nil
 }
 
-func translateToQuayScopes(permission api.Permission) []string {
-	switch permission.Area {
-	case api.PermissionAreaRepository:
-		switch permission.Type {
-		case api.PermissionTypeRead:
-			return []string{"repo:read", "user:read"}
-		case api.PermissionTypeWrite:
-			return []string{"repo:write", "user:read"}
-		case api.PermissionTypeReadWrite:
-			return []string{"repo:read", "repo:write", "user:read"}
+func (p metadataProvider) getEntityRecord(ctx context.Context, token *api.SPIAccessToken, key string, cache map[string]EntityRecord, loginInfoFn func() (LoginTokenInfo, error), fetchFn func(ctx context.Context, cl *http.Client, repoUrl string, tokenData *api.Token, info LoginTokenInfo) (*EntityRecord, error)) (rec EntityRecord, changed bool, err error) {
```

Comment:
The signature of this method is a bit hard to read. Would it be possible to transform somehow parameters?